gadget/install/deploy.go

```diff
@@ -0,0 +1,141 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019-2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package install
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/gadget/internal"
+)
+
+var (
+	deployMountpoint = "/run/snap-recover"
+)
+
+// MakeFilesystem creates a filesystem on the on-disk structure, according
+// to the filesystem type defined in the gadget.
+func MakeFilesystem(ds *gadget.OnDiskStructure) error {
+	if ds.VolumeStructure.HasFilesystem() {
+		fs := ds.VolumeStructure.Filesystem
+		mkfs, ok := internal.MkfsHandlers[fs]
+		if !ok {
+			return fmt.Errorf("cannot create unsupported filesystem %q", fs)
+		}
+		if err := mkfs(ds.Node, ds.VolumeStructure.Label, ""); err != nil {
+			return err
+		}
```

Comment:
With the note I made later, this could be:
```suggestion
		if err := internal.Mkfs(fs, ds.Node, ds.VolumeStructure.Label); err != nil {
			return fmt.Errorf("cannot create filesystem %q: %v", fs, err)
		}
```